x/stake/handler.go

```diff
@@ -53,212 +63,245 @@ func EndBlocker(ctx sdk.Context, k Keeper) (ValidatorUpdates []abci.Validator) {
+	if !found {
+		return ErrBadRedelegationDst(k.Codespace()).Result()
+	}
+	sharesCreated, delegation, dstValidator, pool, err := k.Delegate(ctx, msg.DelegatorAddr, returnCoin, dstValidator)
+	k.SetPool(ctx, pool)
+	k.SetDelegation(ctx, delegation)
+	k.UpdateValidator(ctx, dstValidator)
+
+	// create the unbonding delegation
+	minTime := ctx.BlockHeader().Time + params.UnbondingTime
+	minHeight := ctx.BlockHeight() + params.MinUnbondingBlocks
+
+	red := Redelegation{
+		DelegatorAddr:    msg.DelegatorAddr,
+		ValidatorSrcAddr: msg.ValidatorSrcAddr,
+		ValidatorDstAddr: msg.ValidatorDstAddr,
+		MinTime:          minTime,
+		MinHeight:        minHeight,
+		SharesDst:        sharesCreated,
+		SharesSrc:        msg.SharesAmount,
+	}
+	k.SetRedelegation(ctx, red)
 
-	// remove the bond
-	revokeValidator := false
-	if bond.Shares.IsZero() {
+	tags := sdk.NewTags(
+		tags.Action, tags.ActionBeginRedelegation,
+		tags.Delegator, msg.DelegatorAddr.Bytes(),
+		tags.SrcValidator, msg.ValidatorSrcAddr.Bytes(),
+		tags.DstValidator, msg.ValidatorDstAddr.Bytes(),
+	)
+	return sdk.Result{Tags: tags}
+}
 
-		// if the bond is the owner of the validator then
-		// trigger a revoke validator
-		if bytes.Equal(bond.DelegatorAddr, validator.Owner) &&
-			validator.Revoked == false {
-			revokeValidator = true
-		}
+func handleMsgCompleteRedelegate(ctx sdk.Context, msg types.MsgCompleteRedelegate, k keeper.Keeper) sdk.Result {
 
-		k.removeDelegation(ctx, bond)
-	} else {
-		// Update bond height
-		bond.Height = ctx.BlockHeight()
-		k.setDelegation(ctx, bond)
+	red, found := k.GetRedelegation(ctx, msg.DelegatorAddr, msg.ValidatorSrcAddr, msg.ValidatorDstAddr)
+	if !found {
+		return ErrNoRedelegation(k.Codespace()).Result()
 	}
 
-	// Add the coins
-	pool := k.GetPool(ctx)
-	validator, pool, returnAmount := validator.removeDelShares(pool, delShares)
-	k.setPool(ctx, pool)
-	returnCoins := sdk.Coins{{k.GetParams(ctx).BondDenom, returnAmount}}
-	k.coinKeeper.AddCoins(ctx, bond.DelegatorAddr, returnCoins)
-
-	/////////////////////////////////////
-	// revoke validator if necessary
-	if revokeValidator {
-		validator.Revoked = true
+	// ensure that enough time has passed
+	ctxTime := ctx.BlockHeader().Time
+	ctxHeight := ctx.BlockHeight()
+	if red.MinTime > ctxTime {
+		return ErrNotMature(k.Codespace(), "redelegation", "unit-time", red.MinTime, ctxTime).Result()
 	}
-
-	validator = k.updateValidator(ctx, validator)
-
-	if validator.DelegatorShares.IsZero() {
-		k.removeValidator(ctx, validator.Owner)
+	if red.MinHeight > ctxHeight {
+		return ErrNotMature(k.Codespace(), "redelegation", "block-height", red.MinHeight, ctxHeight).Result()
 	}
 
-	tags := sdk.NewTags("action", []byte("unbond"), "delegator", msg.DelegatorAddr.Bytes(), "validator", msg.ValidatorAddr.Bytes())
-	return sdk.Result{
-		Tags: tags,
-	}
+	k.RemoveRedelegation(ctx, red)
+
+	tags := sdk.NewTags(
+		tags.Action, tags.ActionCompleteRedelegation,
+		tags.Delegator, msg.DelegatorAddr.Bytes(),
```

Comment:
Ditto as above (address encoding in tags) for these three lines

Reply:
addressed